### PR TITLE
feat: update JSON schema and add generated documentation.

### PR DESCRIPTION
**Pull request by jeberhardt**
# Summary | Résumé

There is no issue for this PR.

Making some incremental changes to the JSON schema, and adding some generated documentation.

The documentation was generated via [JSON Schema for Humans](https://github.com/coveooss/json-schema-for-humans). The command to generate the documentation is:

```generate-schema-doc ./lib/middleware/schemas/templates.schema.json ./documentation/Forms/FormViewer.md --config template_name=md```

# Test instructions | Instructions pour tester la modification

-- all existing tests should pass the JSON validation.
-- uploading any JSON configs that meet the JSON schema should pass the validation when uploading.

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

I had to remove:
```        
"properties": {
  "type": "$defs/elementProperties"
}
```
in order to get the tests to pass. I'd like to get this back in, but will add it back in a future PR. This is ongoing work.

Could always add more tests to test against the new schema rules.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [ ] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [ ] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [ ] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: documentation/Forms/FormViewer.md [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+| Type                      | `object`                                                                  |
+| ------------------------- | ------------------------------------------------------------------------- |
+| **Additional properties** | [[Any type: allowed]](# "Additional Properties of any type are allowed.") |
+| **Default**               | `{}`                                                                      |
+
+**Description:** A schema to define a Forms configuration.
+
+| Property                                         | Pattern | Type             | Deprecated | Definition | Title/Description                                                                    |
+| ------------------------------------------------ | ------- | ---------------- | ---------- | ---------- | ------------------------------------------------------------------------------------ |
+| - [securityAttribute](#securityAttribute )       | No      | enum (of string) | No         | -          | It shows an administrator what proceduce should be taken when accessing a form's ... |
```
Comment (Moro-Code) [inline documentation/Forms/FormViewer.md:64]: Finding this a little difficult to read as its cut off

File: documentation/Forms/FormViewer.md [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+| - [internalTitleEn](#internalTitleEn )           | No      | string           | No         | -          | The internal title of the form in English                                            |
+| - [internalTitleFr](#internalTitleFr )           | No      | string           | No         | -          | The internal title of the form in French                                             |
+| + [publishingStatus](#publishingStatus )         | No      | boolean          | No         | -          | If set to false, the link to the form will not appear on the main page, rather u ... |
```
Comment (Moro-Code) [inline documentation/Forms/FormViewer.md:67]: What are the pluses and minuses here supposed to represent ?

File: documentation/Forms/FormViewer.md [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+| + [titleEn](#form_titleEn )               | No      | string          | No         | -          | The displayed title of the form to the user in English                   |
+| + [titleFr](#form_titleFr )               | No      | string          | No         | -          | The displayed title of the form to the user in French                    |
+| - [brand](#form_brand )                   | No      | object          | No         | -          | -                                                                        |
+| - [emailSubjectEn](#form_emailSubjectEn ) | No      | string          | No         | -          | The email subject used in the Notify template in English.                |
+| - [emailSubjectFr](#form_emailSubjectFr ) | No      | string          | No         | -          | The email subject used in the Notify template in French.                 |
+| - [endPage](#form_endPage )               | No      | object          | No         | -          | Content that will be displayed after a form is submitted.                |
+
+### <a name="form_layout"></a>5.1. [Required] Property `GC Forms Schema - Alpha > form > layout`
```
Comment (Moro-Code) [inline documentation/Forms/FormViewer.md:123]: I don't believe the layout is a required property but perhaps I'm mistaken

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Overall I do think it is a great idea to have generated documentation for the JSON structure. However, IMO the MD file is a little difficult to read especially if you do not come from a technical background. Personally ,I myself found it a little difficult to read.  This is just my opinion however and it would be beneficial to get the teams opinion on this 

Furthermore it seems like this requires a python package. If so, instructions should be added to the readme on how to install and use this command.

File: documentation/Forms/FormViewer.md (hunk 3 above)
Comment (jeberhardt) [reply documentation/Forms/FormViewer.md:123]: Not certain if it's required or not either, but in the FormProperties interface, it's not optional.

https://github.com/cds-snc/platform-forms-client/blob/6ec78bba65ece67432236017f571aa8531d3dcb4/lib/types/form-types.ts#L106

File: documentation/Forms/FormViewer.md (hunk 1 above)
Comment (jeberhardt) [reply documentation/Forms/FormViewer.md:64]: It's only cut off in the summary of the attributes. It's not cut off in the section for the attribute itself.

**Comment by jeberhardt** [comment]
> Overall I do think it is a great idea to have generated documentation for the JSON structure. However, IMO the MD file is a little difficult to read especially if you do not come from a technical background. Personally ,I myself found it a little difficult to read. This is just my opinion however and it would be beneficial to get the teams opinion on this

Yeah, I think the template used to create the MD could use a little tweaking to generate documentation that is easier to read. This was just the default MD template, using the default options.

 
> Furthermore it seems like this requires a python package. If so, instructions should be added to the readme on how to install and use this command.

Yep, I can add it to the readme. Maybe also add the command to `package.json`

File: documentation/Forms/FormViewer.md (hunk 2 above)
Comment (jeberhardt) [reply documentation/Forms/FormViewer.md:67]: I'm pretty sure they are meant to represent if an element is required or optional.

File: documentation/Forms/FormViewer.md (hunk 3 above)
Comment (bryan-robitaille) [reply documentation/Forms/FormViewer.md:123]: It's a required property.  The `layout` field is what FormBuilder uses to determine what `elements` are used and in what order they're placed.

**Comment by bryan-robitaille** [comment]
I modified the file locally to see what it would look like when rendered in StoryBook.
I agree with Omar that reducing effort by leveraging generated documentation is nice, from a developer on the project's perspective.  However, if you were a developer on a different project that was sent to that page to integrate with our Template API, would you be as content?
Finding a balance between what we currently have and the dryness of most generated content will most likely be key.

**Comment by jeberhardt** [comment]
Closing this PR until we figure out the format of the generated documentation. Will create another PR with the updates to the schema only.

**Pull request closed without merging**